### PR TITLE
Issue #2166: When decoding the base64-encoded JSON data from backend …

### PR DESCRIPTION
**Pull request by Castaglia**
…Memcached/Redis servers for cached TLS sessions or OCSP responses, perform bounds checking.

### CODE REVIEW

File: contrib/mod_tls_memcache.c [hunk 1]
```diff
@@ -403,22 +403,33 @@ static int sess_cache_entry_decode_json(pool *p, void *value, size_t valuesz,
...
+    return -1;
+  }
 
   key = SESS_CACHE_JSON_KEY_EXPIRES;
   res = entry_get_json_number(p, json, key, &number, entry);
   if (res < 0) {
+    int xerrno = errno;
+
+    (void) pr_json_object_free(json);
+    errno = xerrno;
```
Comment (Castaglia) [inline contrib/mod_tls_memcache.c:416]: Here we are adding better error handling, avoid memory leaks in edge cases, etc.

File: tests/t/lib/ProFTPD/Tests/Modules/mod_tls_memcache.pm [hunk 2]
```diff
@@ -113,6 +117,7 @@ sub tls_sess_cache_memcache {
 
     AuthUserFile => $setup->{auth_user_file},
     AuthGroupFile => $setup->{auth_group_file},
+    AuthOrder => 'mod_auth_file.c',
```
Comment (Castaglia) [inline tests/t/lib/ProFTPD/Tests/Modules/mod_tls_memcache.pm:120]: Here we're dusting off these regression tests, updating them for current test environments, and using them to ensure no egregious regressions.

File: tests/t/modules/mod_tls_redis.t [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+use lib qw(t/lib);
+use strict;
+
+use Test::Unit::HarnessUnit;
+
+$| = 1;
+
+my $r = Test::Unit::HarnessUnit->new();
+$r->start("ProFTPD::Tests::Modules::mod_tls_redis");
```
Comment (Castaglia) [inline tests/t/modules/mod_tls_redis.t:11]: Adding some basic regression tests for the `mod_tls_redis` module; effectively a copy/pasta of the `mod_tls_memcache` regression tests.
